src/thunderbots/software/radio_communication/visitor/mrf_primitive_visitor.cpp

```diff
@@ -0,0 +1,132 @@
+                              " called without ever having visited anything";
+        throw std::runtime_error(err_msg);
+    }
+
+    return *radio_prim;
+}
+
+void MRFPrimitiveVisitor::visit(const CatchPrimitive &catch_primitive)
+{
+    radio_prim              = RadioPrimitive();
+    radio_prim->prim_type   = FirmwarePrimitiveType::CATCH;
+    radio_prim->param_array = {catch_primitive.getVelocity(),
+                               catch_primitive.getDribblerSpeed(),
+                               catch_primitive.getMargin()};
+    radio_prim->extra_bits  = 0;
+}
+
+void MRFPrimitiveVisitor::visit(const ChipPrimitive &chip_primitive)
+{
+    radio_prim              = RadioPrimitive();
+    radio_prim->prim_type   = FirmwarePrimitiveType::SHOOT;
+    radio_prim->param_array = {
+        chip_primitive.getChipOrigin().x() * MILLIMETERS_PER_METER,
+        chip_primitive.getChipOrigin().y() * MILLIMETERS_PER_METER,
+        chip_primitive.getChipDirection().toRadians() * CENTIRADIANS_PER_RADIAN,
+        chip_primitive.getChipDistance() * MILLIMETERS_PER_METER};
+    radio_prim->extra_bits = static_cast<uint8_t>(2 | 1);
+}
+
+void MRFPrimitiveVisitor::visit(const DirectVelocityPrimitive &direct_velocity_primitive)
+{
+    radio_prim              = RadioPrimitive();
+    radio_prim->prim_type   = FirmwarePrimitiveType::DIRECT_VELOCITY;
+    radio_prim->param_array = {
+        direct_velocity_primitive.getXVelocity() * MILLIMETERS_PER_METER,
+        direct_velocity_primitive.getYVelocity() * MILLIMETERS_PER_METER,
+        direct_velocity_primitive.getAngularVelocity() * CENTIRADIANS_PER_RADIAN, 0};
+    radio_prim->extra_bits = direct_velocity_primitive.getDribblerRpm() *
+                             DRIBBLER_RPM_TO_RADIO_CONVERSION_FACTOR;
+}
+
+void MRFPrimitiveVisitor::visit(const DirectWheelsPrimitive &direct_wheels_primitive)
+{
+    radio_prim              = RadioPrimitive();
+    radio_prim->prim_type   = FirmwarePrimitiveType::DIRECT_WHEELS;
+    radio_prim->param_array = {
+        static_cast<double>(direct_wheels_primitive.getWheel0Power()),
+        static_cast<double>(direct_wheels_primitive.getWheel1Power()),
+        static_cast<double>(direct_wheels_primitive.getWheel2Power()),
+        static_cast<double>(direct_wheels_primitive.getWheel3Power())};
+    radio_prim->extra_bits = direct_wheels_primitive.getDribblerRPM() *
+                             DRIBBLER_RPM_TO_RADIO_CONVERSION_FACTOR;
+}
+
+void MRFPrimitiveVisitor::visit(const DribblePrimitive &dribble_primitive)
+{
+    radio_prim              = RadioPrimitive();
+    radio_prim->prim_type   = FirmwarePrimitiveType::DRIBBLE;
+    radio_prim->param_array = {
+        dribble_primitive.getDestination().x() * MILLIMETERS_PER_METER,
+        dribble_primitive.getDestination().y() * MILLIMETERS_PER_METER,
+        dribble_primitive.getFinalAngle().toRadians() * CENTIRADIANS_PER_RADIAN,
+        // There is no division by 300 for RPM in this primitive
```

Comment:
wut. Just out of curiosity, why? also, as a quick suggestion so we don't have values duplicated:
``` suggestion
        // For this primitive, we don't multiply the RPM by anything
```

Reply:
I don't know the answer to this question... was just reading off the firmware implementation of the primitive

Reply:
All good, was just curious